hnix-store-core/src/System/Nix/Internal/StorePath.hs

```diff
@@ -0,0 +1,109 @@
+{-|
+Description : Representation of Nix store paths.
+-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module System.Nix.Internal.StorePath where
+import System.Nix.Hash (HashAlgorithm(Truncated, SHA256), Digest, encodeBase32)
+import Text.Regex.Base.RegexLike (makeRegex, matchTest)
+import Text.Regex.TDFA.Text (Regex)
+import Data.Text (Text)
+import qualified Data.Text as T
+import Data.Text.Encoding (encodeUtf8)
+import GHC.TypeLits (Symbol, KnownSymbol, symbolVal)
+import System.FilePath ((</>))
+import Data.Proxy (Proxy(Proxy))
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BC
+
+-- | The hash algorithm used for store path hashes.
+type StorePathHashAlgo = 'Truncated 20 'SHA256
+
+-- | Regular expression to match valid store path names.
+storePathNameRegex :: Regex
+storePathNameRegex = makeRegex r
+  where
+    r :: String
+    r = "[a-zA-Z0-9\\+\\-\\_\\?\\=][a-zA-Z0-9\\+\\-\\.\\_\\?\\=]*"
+
+-- | The name portion of a Nix path.
+--
+-- 'unStorePathName' must only contain a-zA-Z0-9+._?=-, can't start
+-- with a -, and must have at least one character (i.e. it must match
+-- 'storePathNameRegex').
+newtype StorePathName = StorePathName
+  { -- | Extract the contents of the name.
+    unStorePathName :: Text
+  }
+
+-- | Smart constructor for 'StorePathName' that ensures the underlying
+-- content invariant is met.
+makeStorePathName :: Text -> Maybe StorePathName
+makeStorePathName n = case matchTest storePathNameRegex n of
+  True  -> Just $ StorePathName n
+  False -> Nothing
+
+-- | A type-level representation of the root directory of a Nix store.
+type StoreDir = Symbol
```

Comment:
Does `StoreDir` really need to be at the type level?  I'm not seeing the value.  Why not keep it as a normal value-level field in `StorePath`?

Reply:
I'll update the comment with more details, but just a tidbit: There are 30 places in the C++ nix codebase where we have a runtime assertion that is only there because the type isn't strong enough :wink:

Reply:
Is there a way to take some arbitrary storeDir CLI argument and make a `KnownSymbol` from it?

Reply:
http://hackage.haskell.org/package/base-4.12.0.0/docs/GHC-TypeLits.html#v:someSymbolVal

Reply:
Added detailed docs explaining this decision.